src/shogun/kernel/SpectralMixtureKernel.cpp

```diff
@@ -0,0 +1,221 @@
+
+	return result;
+}
+
+SGVector<float64_t> CSpectralMixtureKernel::get_squared_features(CDotFeatures* features)
+{
+	SGVector<float64_t> sqr_features = SGVector<float64_t>(features->get_num_vectors());
+	
+	for (int i=0; i<sqr_features.size(); i++)
+		sqr_features[i] = features->dot(i, features, i);
+
+	return sqr_features;
+}
+
+SGMatrix<float64_t> CSpectralMixtureKernel::get_square_distance_matrix()
+{
+	CDotFeatures* dot_lhs = dynamic_cast<CDotFeatures*>(lhs);
+	CDotFeatures* dot_rhs = dynamic_cast<CDotFeatures*>(rhs);
+
+	REQUIRE(dot_lhs != NULL, "Left-hand-side features must be of type CDotFeatures");
+	REQUIRE(dot_rhs != NULL, "Right-hand-side features must be of type CDotFeatures");
+	
+	SGMatrix<float64_t> lhs_mtrx = dot_lhs->get_computed_dot_feature_matrix();
+	SGMatrix<float64_t> rhs_mtrx = dot_rhs->get_computed_dot_feature_matrix();
+
+	/* rhs^T . lhs: index (i, j) is dot product of lhs_i and rhs_j.
+	   result scaled by minus two for distance computation */
+	SGMatrix<float64_t> dot_mtrx = SGMatrix<float64_t>::matrix_multiply(rhs_mtrx, lhs_mtrx, true, false, -2.0);
+
+	SGVector<float64_t> sqr_lhs = get_squared_features(dot_lhs);
+	SGVector<float64_t> sqr_rhs = get_squared_features(dot_rhs);
+
+	for (int i=0; i<dot_mtrx.num_cols; i++)
+	{
+		/* To each column vector is added a scalar: lhs_i^2
+		   and the vector [rhs_1^2 ... rhs_n^2]^T*/
+		float64_t* col_vector_ptr = dot_mtrx.get_column_vector(i);
+
+		SGVector<float64_t>::add_scalar(sqr_lhs[i], col_vector_ptr, dot_mtrx.num_rows);
+		SGVector<float64_t>::add(col_vector_ptr, 1.0, col_vector_ptr, 1.0, sqr_rhs, dot_mtrx.num_rows);
+	}
+	
+	return dot_mtrx;
+}
+
+void CSpectralMixtureKernel::compute_weight_derivative(SGMatrix<float64_t>& squared_distances, index_t index)
+{
+	for (int i=0; i<squared_distances.num_rows; i++)
+	{
+		for (int j=0; j<squared_distances.num_cols; j++)
+		{
+			// 2*w_i*exp(-2*pi^2*dist^2*sigma_i^2) * cos(2*pi*dist*mu_i)
+			float64_t sqr_dist = squared_distances(i,j);
+			float64_t exp_factor = CMath::exp(-2*CMath::pow(CMath::PI,2)*sqr_dist*CMath::pow(m_stddeviations[index],2));
+			float64_t cos_factor = CMath::cos(2*CMath::PI*CMath::sqrt(sqr_dist)*m_means[index]);
+			squared_distances(i, j) = 2*m_weights[index]*exp_factor*cos_factor;
+		}
+	}
+}
+
+void CSpectralMixtureKernel::compute_mean_derivative(SGMatrix<float64_t>& squared_distances, index_t index)
+{
+	for (int i=0; i<squared_distances.num_rows; i++)
+	{
+		for (int j=0; j<squared_distances.num_cols; j++)
+		{
+			// -2*pi*dist*mu_i^2*exp(-2*pi^2*dist^2*sigma_i^2)*sin(2*pi*dist*mu_i)
+			float64_t sqr_dist = squared_distances(i,j);
+			float64_t outer_factor = -2*CMath::PI*CMath::sqrt(sqr_dist)*CMath::pow(m_weights[index],2);
+			float64_t exp_factor = CMath::exp(-2*CMath::pow(CMath::PI, 2)*sqr_dist*CMath::pow(m_stddeviations[index], 2));
```

Comment:
Benchmark result based on [this](https://gist.github.com/lambday/e5234e2e564e6af637a7) (100x100 matrices), element-wise `sin` operation
```
[lambday@lambday.iitb.ac.in benchmarks]$ ./benchmark 
[==========] Running 3 benchmarks.
[ RUN      ] SGMatrix.elementwise_sin_NATIVE (10 runs, 1000 iterations per run)
[     DONE ] SGMatrix.elementwise_sin_NATIVE (3489.504000 ms)
[   RUNS   ]        Average time: 348950.400 us
                         Fastest: 342721.000 us (-6229.400 us / -1.785 %)
                         Slowest: 371872.000 us (+22921.600 us / +6.569 %)
                                  
             Average performance: 2.86574 runs/s
                Best performance: 2.91783 runs/s (+0.05209 runs/s / +1.81763 %)
               Worst performance: 2.68910 runs/s (-0.17664 runs/s / -6.16384 %)
[ITERATIONS]        Average time: 348.950 us
                         Fastest: 342.721 us (-6.229 us / -1.785 %)
                         Slowest: 371.872 us (+22.922 us / +6.569 %)
                                  
             Average performance: 2865.73679 iterations/s
                Best performance: 2917.82529 iterations/s (+52.08849 iterations/s / +1.81763 %)
               Worst performance: 2689.09732 iterations/s (-176.63947 iterations/s / -6.16384 %)
[ RUN      ] SGMatrix.elementwise_sin_EIGEN3 (10 runs, 1000 iterations per run)
[     DONE ] SGMatrix.elementwise_sin_EIGEN3 (3592.366000 ms)
[   RUNS   ]        Average time: 359236.600 us
                         Fastest: 357530.000 us (-1706.600 us / -0.475 %)
                         Slowest: 363527.000 us (+4290.400 us / +1.194 %)
                                  
             Average performance: 2.78368 runs/s
                Best performance: 2.79697 runs/s (+0.01329 runs/s / +0.47733 %)
               Worst performance: 2.75083 runs/s (-0.03285 runs/s / -1.18021 %)
[ITERATIONS]        Average time: 359.237 us
                         Fastest: 357.530 us (-1.707 us / -0.475 %)
                         Slowest: 363.527 us (+4.290 us / +1.194 %)
                                  
             Average performance: 2783.68073 iterations/s
                Best performance: 2796.96809 iterations/s (+13.28736 iterations/s / +0.47733 %)
               Worst performance: 2750.82731 iterations/s (-32.85342 iterations/s / -1.18021 %)
[ RUN      ] CGPUMatrix.elementwise_sin_VIENNACL (10 runs, 1000 iterations per run)
[     DONE ] CGPUMatrix.elementwise_sin_VIENNACL (129.061000 ms)
[   RUNS   ]        Average time: 12906.100 us
                         Fastest: 4283.000 us (-8623.100 us / -66.814 %)
                         Slowest: 55755.000 us (+42848.900 us / +332.005 %)
                                  
             Average performance: 77.48274 runs/s
                Best performance: 233.48120 runs/s (+155.99846 runs/s / +201.33318 %)
               Worst performance: 17.93561 runs/s (-59.54713 runs/s / -76.85212 %)
[ITERATIONS]        Average time: 12.906 us
                         Fastest: 4.283 us (-8.623 us / -66.814 %)
                         Slowest: 55.755 us (+42.849 us / +332.005 %)
                                  
             Average performance: 77482.74072 iterations/s
                Best performance: 233481.20476 iterations/s (+155998.46404 iterations/s / +201.33318 %)
               Worst performance: 17935.61116 iterations/s (-59547.12956 iterations/s / -76.85212 %)
[==========] Ran 3 benchmarks.
```
GPU is way faster! Will try with custom unary operator (as required for this PR) as well.